ipsec/sadb.c

```diff
@@ -0,0 +1,926 @@
+}
+
+static int sadb_x_spddel(const struct sadb_msg *msg, struct sadb_msg *reply)
+{
+	union sadb_headers h;
+	spd_table *spd;
+	spd_entry_t *sp;
+	db_set_netif *db_sets;
+	int ret;
+
+	if (msg == NULL || reply == NULL)
+		return -EIO;
+
+	_sadb_deserializeMsg(msg, &h);
+
+	if (h.policy == NULL)
+		return -EINVAL;
+
+	if (h.policy->sadb_x_policy_id == 0)
+		return -EINVAL;
+
+	db_sets = ipsecdev_dbsget();
+	if (db_sets == NULL)
+		return -EINVAL;
+
+	if (h.policy->sadb_x_policy_dir == IPSEC_DIR_OUTBOUND)
+		spd = &db_sets->outbound_spd;
+	else if (h.policy->sadb_x_policy_dir == IPSEC_DIR_INBOUND)
+		spd = &db_sets->inbound_spd;
+	else
+		return -EINVAL;
+
+	sp = (void *)h.policy->sadb_x_policy_id;
+	ret = ipsec_spd_del_maybe(sp, spd);
+
+	reply->sadb_msg_len = PFKEY_UNIT64(_sadb_serializeMsg(reply, &h));
+	return ret;
+}
+
+static void ipsec_dump_tables(int dump_spd)
+{
+	db_set_netif *db_sets;
+
+	db_sets = ipsecdev_dbsget();
+	if (db_sets == NULL)
+		return;
+
+	if (dump_spd) {
+		ipsec_spd_dump_log(&db_sets->inbound_spd, "In  ");
+		ipsec_spd_dump_log(&db_sets->outbound_spd, "Out ");
+	}
+	else {
+		ipsec_sad_dump_log(&db_sets->inbound_sad, "In  ");
+		ipsec_sad_dump_log(&db_sets->outbound_sad, "Out ");
+	}
+}
+
+#define CHECK_TIMEOUT_SEC 5
+
+static void *ipsec_check_timeouts(void *arg)
+{
+	db_set_netif *db_sets;
+	sad_entry_t *sa_expired = NULL;
+	int is_soft;
+	char msgbuf[512];
+	struct sadb_msg *msg = (struct sadb_msg *)msgbuf;
+
+	db_sets = ipsecdev_dbsget();
+	if (db_sets == NULL)
+		return NULL;
+
+	while (1) {
+		do {
+			sleep(CHECK_TIMEOUT_SEC);
+		} while (!sadb_check_timeouts_enabled);
+
+		// NOTE: we're sending one expiration at a time to avoid loops (the user is responsible for sending SADB_DELETE)
+		if (((sa_expired = ipsec_sad_check_timeouts(&db_sets->inbound_sad, &is_soft)) != NULL) || ((sa_expired = ipsec_sad_check_timeouts(&db_sets->outbound_sad, &is_soft)) != NULL)) {
+			msg->sadb_msg_len = PFKEY_UNIT64(sizeof(msgbuf));
+			sadb_fill_msg_common(msg, sa_expired, (is_soft ? 1 : 2));
+			msg->sadb_msg_type = SADB_EXPIRE;
+			int is_larval = sa_expired->initiator != 0;
+
+			IPSEC_LOG_SADB("SADB_EXPIRE (%s): spi: %x  mode: %d  dst_addr: %x",
+				(is_soft ? "soft" : "hard"), sa_expired->spi, sa_expired->mode, sa_expired->addr.addr);
+
+			key_sockets_notify(msg);
+
+			if (!is_soft) {
+				// note: we should mark SA not to be used again
+			}
+
+			if (!is_soft && is_larval) {
+				// NOTE: i don't know if this is legal, but iked never deletes LARVAL spi's
+				ipsec_sad_del_spi(sa_expired->spi, &db_sets->outbound_sad);
+				ipsec_sad_del_spi(sa_expired->spi, &db_sets->inbound_sad);
+			}
+		}
+	}
+
+	return NULL;
+}
+
+void ipsec_sadbStartCheckingTimeouts(void)
+{
+	sadb_check_timeouts_enabled = 1;
+}
+
+void ipsec_sadbStopCheckingTimeouts(void)
+{
+	sadb_check_timeouts_enabled = 0;
+}
+
+void ipsec_sadbInitCheckingTimeouts(void)
+{
+	pthread_t thread;
+
+	pthread_create(&thread, NULL, ipsec_check_timeouts, NULL);
```

Comment:
In LWIP we need to use it's mini-lib for threads handling.
https://github.com/phoenix-rtos/phoenix-rtos-lwip/blob/master/port/threads.c